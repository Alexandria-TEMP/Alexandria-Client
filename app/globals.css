@tailwind base;
@tailwind components;
@tailwind utilities;

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}

@layer base {
  h1 {
    @apply text-2xl font-bold;
  }
  h2 {
    @apply text-xl font-semibold;
  }
  h3 {
    @apply text-lg;
  }
  h6 {
    @apply text-sm;
  }
}
