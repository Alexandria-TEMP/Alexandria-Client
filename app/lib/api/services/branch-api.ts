import { BranchUnionT, idBranchUnionT } from "@/lib/types/branch-union";
import { BranchT, ClosedBranchT, idT } from "../../types/api-types";
import { baseUrl, validateResponse } from "../api-common";
import fetchPostData from "./post-api";

/**
 * Fetches branch or closed branch data in a unified object
 * @async
 * @param id branch or closed branch ID
 * @returns branch and optionally closed branch data
 */
export async function fetchBranchData(
  id: idBranchUnionT,
): Promise<BranchUnionT> {
  let closedBranch = undefined;

  if (id.isClosed) {
    const closedBranchResponse = await fetch(
      `${baseUrl}/branches/closed/${id.id}`,
    );
    await validateResponse(closedBranchResponse);
    closedBranch = (await closedBranchResponse.json()) as ClosedBranchT;
  }

  const branchID = closedBranch?.branchID ?? (id.id as idT);
  const branchResponse = await fetch(`${baseUrl}/branches/${branchID}`);
  await validateResponse(branchResponse);

  const branch = (await branchResponse.json()) as BranchT;

  const projectPostID = branch.projectPostID ?? closedBranch?.projectPostID;
  if (!projectPostID) {
    throw new Error(
      `missing some project post ID in ${id.isClosed ? "closed" : ""} branch ${id.id}`,
    );
  }

  const updated = await fetchBranchUpdatedFieldsFallback(branch, projectPostID);

  return { branch, closedBranch, updated, projectPostID };
}

/**
 * For each of the possible post fields that a branch updates, returns
 * either the updated data or the current data if updated is null
 * @param branch branch whose update we're interested in
 * @param projectPostID ID of project post that branch is updating
 */
export async function fetchBranchUpdatedFieldsFallback(
  branch: BranchT,
  projectPostID: idT,
) {
  if (
    branch.updatedPostTitle &&
    branch.updatedCompletionStatus &&
    branch.updatedScientificFieldTagContainerID
  ) {
    return {
      postTitle: branch.updatedPostTitle,
      completionStatus: branch.updatedCompletionStatus,
      scientificFieldTagContainerID:
        branch.updatedScientificFieldTagContainerID,
    };
  }

  const postData = await fetchPostData({
    id: projectPostID,
    isProject: true,
  });

  return {
    postTitle: branch.updatedPostTitle ?? postData.post.title,
    completionStatus:
      branch.updatedCompletionStatus ??
      postData.projectPost!.projectCompletionStatus,
    scientificFieldTagContainerID:
      branch.updatedScientificFieldTagContainerID ??
      postData.post.scientificFieldTagContainerID,
  };
}

/**
 * Gets branches of a post given their ID.
 * @async
 * @param id Post ID
 */
// TODO remove next line
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export async function getPostBranches(id: idT) {
  await new Promise((resolve) => setTimeout(resolve, 100));
  // TODO
  return {
    rejected: [1],
    open: [2],
    accepted: [3],
  };
}

/**
 * TODO jsdoc when properly implemented
 */
// TODO remove next line
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export async function getBranchReviewStatuses(id: idT) {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 70));
  return [];
  // if (id == 1) return ["accept", "reject", "accept"];
  // else if (id == 2) return ["accept", "open", "open"];
  // else return ["accept", "accept", "accept"];
}
