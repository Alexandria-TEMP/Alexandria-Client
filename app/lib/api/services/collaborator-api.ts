/* eslint-disable @typescript-eslint/no-unused-vars */
// TODO remove previous line once implemented

import { idT } from "../../types/api-types";
import { QuartoContainerTypeT } from "../../types/quarto-container";
import { baseUrl } from "../api-common";

/**
 * Builds URL path for collaborator API calls
 * @param id collaborator ID
 * @param containerType if Quarto project is in a post or branch
 */
const buildResourcePath = (id: idT, containerType: QuartoContainerTypeT) =>
  `${baseUrl}/${containerType === "branch" ? "branches" : "posts"}/collaborators/${id}`;

/**
 * TODO
 */
export async function fetchCollaborators(
  ids: idT[],
  containerType: QuartoContainerTypeT,
) {
  // TODO
}
