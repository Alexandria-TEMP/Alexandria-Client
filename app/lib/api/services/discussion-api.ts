import { DiscussionContainerT, idT } from "@/lib/types/api-types";
import { baseUrl, validateResponse } from "../api-common";

/**
 * Get a discussion container by its ID, to access its discussions
 * @param id discussion container ID
 */
export async function fetchDiscussionContainer(
  id: idT,
): Promise<DiscussionContainerT> {
  const res = await fetch(`${baseUrl}/discussion-containers/${id}`);
  await validateResponse(res);
  return (await res.json()) as DiscussionContainerT;
}

/**
 * Gets data for a discussion given its ID.
 * @async
 * @param id Discussion ID
 */
export async function getDiscussionData(id: string) {
  // TODO
  // const res = await fetch();
  await new Promise((resolve) => setTimeout(resolve, 100));

  return {
    id: id,
    anonymous: false,
    author: {
      email: "mariecurie@tudelft.nl",
      firstName: "Marie",
      institution: "TU Delft",
      lastName: "Curie",
    },
    createdAt: "11 May 2024",
    deleted: false,
    deletedAt: "-",
    replies: [],
    text: "Lorem ipsum dolor sit amet, consectetur adipisicing elit. Sunt ut ducimus amet ex qui eius corrupti reiciendis, quibusdam suscipit, aspernatur ipsum. Reprehenderit libero molestias nostrum eum sed? Illo, quidem ad.",
    updatedAt: "11 May 2024",
  };
}

// TODO parameter type should include all data needed for a discussion
/**
 * Creates a new Discussion for a given Version.
 * @async
 * @param text Discussion contents
 * @param id discussion container ID
 */
export async function uploadDiscussion(text: string, id: idT) {
  // TODO
  // should include author, possible different discussion to reply to (and anything else that I'm forgetting)
  await new Promise((resolve) => setTimeout(resolve, 100));
  return `uploaded ${text} under id ${id}`;
}
