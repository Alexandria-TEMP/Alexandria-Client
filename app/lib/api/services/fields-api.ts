import {
  ScientificFieldTagContainerT,
  ScientificFieldTagT,
  idT,
} from "../../types/api-types";
import { baseUrl, validateResponse } from "../api-common";

/**
 * TODO jsdoc when properly implemented
 */
export async function getFields() {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 1000));
  const data = [
    {
      id: "1",
      tag: "Computer Science",
      tagType: "ScientificField",
    },
    {
      id: "2",
      tag: "Mathematics",
      tagType: "ScientificField",
    },
    {
      id: "3",
      tag: "Psychology & Psychiatry",
      tagType: "ScientificField",
    },
    {
      id: "4",
      tag: "Medicine",
      tagType: "ScientificField",
    },
  ];

  return data;
}

/**
 * Fetches all scientific fields based on their IDs
 * @param ids scientific field IDs
 */
export async function fetchScientificFields(
  ids: idT[],
): Promise<ScientificFieldTagT[]> {
  let tags: ScientificFieldTagT[] = [];

  for (const id of ids) {
    const res = await fetch(`${baseUrl}/tags/scientific/${id}`);
    if (!res.ok) continue;
    const scientificField = (await res.json()) as ScientificFieldTagT;
    tags = [...tags, scientificField];
  }

  return tags;
}

/**
 * Fetch scientific field tag container
 * @param id of the container
 * @returns container object
 */
export async function fetchScientificFieldContainer(
  id: idT,
): Promise<ScientificFieldTagContainerT> {
  const res = await fetch(`${baseUrl}/tags/scientific/containers/${id}`);
  await validateResponse(res);
  const container = (await res.json()) as ScientificFieldTagContainerT;
  return container;
}

/**
 * Method that gets the scientific fiels from a container
 * @param id container id
 * @returns a list of scientific field tags in that container
 */
export async function fetchScientificFieldsFromContainer(
  id: idT,
): Promise<ScientificFieldTagT[]> {
  const container = await fetchScientificFieldContainer(id);
  return await fetchScientificFields(container.scientificFieldTagIDs);
}
