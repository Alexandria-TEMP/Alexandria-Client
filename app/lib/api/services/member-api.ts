// methods in this file will be heavily changed once integration with back end it done
// just retrieve some dummy data for now

import { MemberT, idT } from "../../types/api-types";

/**
 * Gets data for a Member given their ID.
 * @async
 * @param id Member ID
 */
export default async function getMemberData(id: idT): Promise<MemberT> {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 100));
  if (id == 1) {
    return {
      id,
      email: "mariecurie@tudelft.nl",
      firstName: "Marie",
      institution: "TU Delft",
      lastName: "Curie",
      scientificFieldTagIDs: [],
    };
  } else if (id == 2) {
    return {
      id,
      email: "kopernicus@tudelft.nl",
      firstName: "Nicolaus",
      institution: "TU Delft",
      lastName: "Copernicus",
      scientificFieldTagIDs: [],
    };
  } else if (id == 3) {
    return {
      id,
      email: "kopernicus@tudelft.nl",
      firstName: "Metal Bar",
      institution: "TU Delft",
      lastName: "Clanging",
      scientificFieldTagIDs: [],
    };
  } else {
    return {
      id,
      email: "kopernicus@tudelft.nl",
      firstName: "Michael",
      institution: "TU Delft",
      lastName: "Yippie",
      scientificFieldTagIDs: [],
    };
  }
}

/**
 * Method that gets all members from the database, should be deprecated?
 * @returns A map of all members with their database id as key, and the whole member object as value
 */
export async function getMembers() {
  // TODO implement
  // TODO edit jsdoc (deprecation remark)
  // pretend this does multiple fetches
  await new Promise((resolve) => setTimeout(resolve, 100));
  const members = [
    {
      id: "1",
      email: "mariecurie@tudelft.nl",
      firstName: "Marie",
      picture: "/placeholders/Marie_Curie.jpg",
      institution: "TU Delft",
      lastName: "Curie",
    },
    {
      id: "2",
      email: "kopernicus@tudelft.nl",
      firstName: "Nicolaus",
      institution: "TU Delft",
      picture: "/placeholders/Nikolaus_Kopernikus.jpg",
      lastName: "Copernicus",
    },
    {
      id: "3",
      email: "kopernicus@tudelft.nl",
      firstName: "Metal Bar",
      institution: "TU Delft",
      picture: "/placeholders/Nikolaus_Kopernikus.jpg",
      lastName: "Clanging",
    },
    {
      id: "4",
      email: "kopernicus@tudelft.nl",
      firstName: "Michael",
      institution: "TU Delft",
      picture: "/placeholders/Nikolaus_Kopernikus.jpg",
      lastName: "Yippie",
    },
  ];

  return members;
}
