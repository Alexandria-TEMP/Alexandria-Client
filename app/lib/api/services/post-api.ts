import { PostT, ProjectPostT, idT } from "../../types/api-types";
import { PostUnionT, idPostUnionT } from "../../types/post-union";
import { baseUrl, validateResponse } from "../api-common";

/**
 * Fetches post or project post data in a unified object
 * @param id post or project post id
 * @returns post and optionally project post data
 */
export default async function fetchPostData(
  id: idPostUnionT,
): Promise<PostUnionT> {
  let projectPost = undefined;

  if (id.isProject) {
    const projectPostResponse = await fetch(
      `${baseUrl}/project-posts/${id.id}`,
    );
    await validateResponse(projectPostResponse);
    projectPost = (await projectPostResponse.json()) as ProjectPostT;
  }

  const postID = projectPost?.postID ?? (id.id as idT);
  const postResponse = await fetch(`${baseUrl}/posts/${postID}`);
  await validateResponse(postResponse);

  const post = (await postResponse.json()) as PostT;

  return { post, projectPost };
}
