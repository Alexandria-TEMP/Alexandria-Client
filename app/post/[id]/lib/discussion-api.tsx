/**
 * Gets data for a discussion given its ID.
 *
 * @async
 * @param id Discussion ID
 */
export async function getDiscussionData(id: string) {
  // TODO
  // const res = await fetch();

  return {
    anonymous: false,
    author: {
      email: "mariecurie@tudelft.nl",
      firstName: "Marie",
      institution: "TU Delft",
      lastName: "Curie",
    },
    createdAt: "11 May 2024",
    deleted: false,
    deletedAt: "-",
    replies: [],
    text: "Lorem ipsum dolor sit amet, consectetur adipisicing elit. Sunt ut ducimus amet ex qui eius corrupti reiciendis, quibusdam suscipit, aspernatur ipsum. Reprehenderit libero molestias nostrum eum sed? Illo, quidem ad.",
    updatedAt: "11 May 2024",
  };
}

// TODO parameter type should include all data needed for a discussion
/**
 * Creates a new Discussion for a given Version.
 *
 * @async
 * @param text Discussion contents
 * @param id ID of the Version the Discussion references
 */
export async function uploadDiscussion(text: string, id: string) {
  // TODO
  // should include author, possible different discussion to reply to (and anything else that I'm forgetting)
  return Promise.resolve();
}
