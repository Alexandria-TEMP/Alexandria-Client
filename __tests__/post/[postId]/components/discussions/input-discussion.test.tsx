import { expect, describe, it } from "@jest/globals";
import userEvent from "@testing-library/user-event";
import { render, screen } from "@testing-library/react";
import { uploadDiscussion } from "@/lib/api/services/discussion-api";
import InputDiscussion from "@/post/[postId]/components/discussions/input-discussion";

jest.mock("@/lib/api/services/discussion-api");

describe("InputDiscussion", () => {
  const discussionContainerID = 51235312;
  const mockedApi = uploadDiscussion as jest.Mock;
  mockedApi.mockResolvedValue({});

  it("renders a textbox", () => {
    render(<InputDiscussion id={discussionContainerID} />);
    expect(screen.getByRole("textbox")).toBeInTheDocument();
  });

  it("calls uploadDiscussion with input", async () => {
    const user = userEvent.setup();
    const inputText = "Alexandria is pretty cool";

    render(<InputDiscussion id={discussionContainerID} />);

    await user.type(screen.getByRole("textbox"), inputText);
    await user.click(screen.getByRole("button"));

    expect(mockedApi).toHaveBeenCalledWith(inputText, discussionContainerID);
  });
});
