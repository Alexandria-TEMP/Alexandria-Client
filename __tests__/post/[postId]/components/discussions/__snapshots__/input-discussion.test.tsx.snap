// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputDiscussion matches snapshot normal 1`] = `
<div>
  <form>
    <div
      class="group flex flex-col w-full"
      data-filled="true"
      data-filled-within="true"
      data-has-elements="true"
      data-has-label="true"
      data-slot="base"
    >
      <label
        class="z-10 pointer-events-none origin-top-left rtl:origin-top-right subpixel-antialiased block text-small text-foreground-500 relative will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] group-data-[filled-within=true]:text-foreground group-data-[filled-within=true]:pointer-events-auto pb-1.5 pe-2 max-w-full text-ellipsis overflow-hidden"
        data-slot="label"
        for="react-aria-:r0:"
        id="react-aria-:r1:"
      >
        <h2>
          Your reply
        </h2>
      </label>
      <div
        class="relative w-full inline-flex tap-highlight-transparent flex-row items-center shadow-sm px-3 gap-3 bg-default-100 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 h-10 min-h-10 rounded-medium !h-auto transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background py-2"
        data-has-multiple-rows="true"
        data-slot="input-wrapper"
        style="cursor: text;"
      >
        <div
          class="inline-flex w-full h-full box-border items-start group-data-[has-label=true]:items-start"
          data-slot="inner-wrapper"
        >
          <textarea
            aria-label="Start a new discussion..."
            aria-labelledby="react-aria-:r0: react-aria-:r1:"
            class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 text-small resize-none data-[hide-scroll=true]:scrollbar-hide group-data-[has-value=true]:text-default-foreground h-full transition-height !duration-100 motion-reduce:transition-none"
            data-hide-scroll="true"
            data-slot="input"
            data-testid="input-discussion-textarea"
            id="react-aria-:r0:"
            name="text"
            placeholder="Start a new discussion..."
            title=""
          />
        </div>
      </div>
    </div>
    <div
      class="flex flex-row mt-4 content-center gap-2"
    >
      <label
        class="group relative max-w-fit inline-flex items-center justify-start cursor-pointer touch-none tap-highlight-transparent"
      >
        <div
          style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
        >
          <input
            aria-labelledby=":r5:"
            role="switch"
            type="checkbox"
            value=""
          />
        </div>
        <span
          aria-hidden="true"
          class="px-1 relative inline-flex items-center justify-start flex-shrink-0 overflow-hidden bg-default-200 rounded-full outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background group-data-[selected=true]:bg-primary group-data-[selected=true]:text-primary-foreground w-10 h-6 mr-2 rtl:ml-2 rtl:mr-[unset] transition-background"
        >
          <span
            class="z-10 flex items-center justify-center bg-white shadow-small rounded-full origin-right w-4 h-4 text-tiny group-data-[selected=true]:ml-4 rtl:group-data-[selected=true]:ml-0 rtl:group-data-[selected=true]:mr-4 transition-all group-data-[pressed=true]:w-5 group-data-[selected]:group-data-[pressed]:ml-3"
          />
        </span>
        <span
          class="relative text-foreground select-none text-small"
          id=":r5:"
        >
          <h6
            class="text-neutral-400"
          >
            Anonymous reply
          </h6>
           
        </span>
      </label>
      <div
        class="grow"
      />
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] data-[pressed=true]:scale-[0.97] transition-transform-colors-opacity motion-reduce:transition-none bg-default text-default-foreground data-[hover=true]:opacity-hover"
        data-testid="submit-discussion"
        type="submit"
      >
        Post your answer
      </button>
    </div>
  </form>
</div>
`;
